### ISSUE TITLE
Question regarding the logits in the `_get_batch_logps` function

### ISSUE THREAD

**Issue by vgoklani** [eric-mitchell/direct-preference-optimization#61]
For the log probability calculation:

    labels = labels[:, 1:].clone()
    logits = logits[:, :-1, :]

Why do we use logits[:, :-1, :] instead of logits[:, 1:, :] ?

https://github.com/eric-mitchell/direct-preference-optimization/blob/main/trainers.py#L104

Thanks!

**Comment by rustic-snob** [comment]
Because it is to calculated the rolled-out output. the logit should always predicting the next token in sequence length dimension.

It is almost the same with transformers ~ModelForCausalLM.
https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py#L1205C1-L1205C1

**Comment by eric-mitchell** [comment]
`logits[i,j,:]` are the predicted unnormalized (pre-softmax) scores for the next token, i.e., the prediction for the token index in `labels[i,j+1]`. That's why we shift them relative to each other.
